cudaaligner/src/aligner_global_myers_banded.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "aligner_global_myers_banded.hpp"
+#include "myers_gpu.cuh"
+#include "batched_device_matrices.cuh"
+
+#include <claragenomics/utils/mathutils.hpp>
+
+namespace claragenomics
+{
+
+namespace cudaaligner
+{
+
+struct AlignerGlobalMyersBanded::Workspace
+{
+    Workspace(int32_t max_alignments, int32_t max_n_words, int32_t max_target_length, cudaStream_t stream)
+        : pvs(max_alignments, max_n_words * (max_target_length + 1), stream)
+        , mvs(max_alignments, max_n_words * (max_target_length + 1), stream)
+        , scores(max_alignments, max_n_words * (max_target_length + 1), stream)
+        , query_patterns(max_alignments, max_n_words * 4, stream)
+    {
+    }
+    batched_device_matrices<myers::WordType> pvs;
+    batched_device_matrices<myers::WordType> mvs;
+    batched_device_matrices<int32_t> scores;
+    batched_device_matrices<myers::WordType> query_patterns;
+};
+
+AlignerGlobalMyersBanded::AlignerGlobalMyersBanded(int32_t max_query_length, int32_t max_target_length, int32_t max_alignments, cudaStream_t stream, int32_t device_id)
+    : AlignerGlobal(max_query_length, max_target_length, max_alignments, stream, device_id)
+    , workspace_()
+{
+    scoped_device_switch dev(device_id);
+    workspace_ = std::make_unique<Workspace>(max_alignments, ceiling_divide<int32_t>(max_query_length, sizeof(myers::WordType)), max_target_length, stream);
+}
+
+AlignerGlobalMyersBanded::~AlignerGlobalMyersBanded()
+{
+    // Keep empty destructor to keep Workspace type incomplete in the .hpp file.
```

Comment:
My knowledge gap is showing here - can you explain this a bit more for me?

Reply:
AlignerGlobalMyersBanded has a `unique_ptr<Workspace>`, which means the AlignerGlobalMyersBanded's dtor has to call delete on `Workspace`.
Now `Workspace` is defined in the .cpp file only. In the .hpp file, `Workspace` is only declared as "some" struct, so the compiler does not know the size of the Workspace object by looking at the .hpp file (nor does it know if it has to call destructors for data members). Therefore, it cannot delete it if it only has the .hpp file, but has to go to the .cpp file.

If I write `~AlignerGlobalMyersBanded = default;` in the .hpp file, it will generate a destructor inline (for all files that include the .hpp file), which will then try to delete Workspace inline. This fails for the reason mentioned above.

So the solution is to prevent the compiler from generating a destructor inline, but to explicitly tell the compiler there is a destructor in the .cpp file and it has to call that. In the .cpp file the compiler can then generate the delete Workspace, since it sees its definition.

Reply:
Cool, thanks for the explanation!